sycl/doc/design/DeviceGlobal.md

```diff
@@ -563,11 +563,20 @@ instance of a device global variable in a `pi_program`.  This functionality is
 exposed as two new PI interfaces:
 
 ```
-pi_result piextCopyToDeviceVariable(pi_device Device, pi_program Program,
-  const char *name, const void *src, size_t count, size_t offset);
-
-pi_result piextCopyFromDeviceVariable(pi_device Device, pi_program Program,
-  const char *name, void *dst, size_t count, size_t offset);
+pi_result piextEnqueueDeviceVariableRead(pi_queue Queue, pi_program Program,
+                                         const char *Name, pi_bool BlockingRead,
```

Comment:
> Can we instead add an API to retrieve variable's location from its Name and use existing memory reading/writing API on that?

I wouldn't mind that as it seems like that's how both CUDA and L0 handles it, however it would require a change to `opencl-extensions/cl_intel_global_variable_access.asciidoc`.

> Also how is the Count size known to the caller? And what is the semantics of the Offset?

The offset is into the device variable, i.e. the operation will copy `Count` bytes to/from the device variable at byte-offset `Offset`.

Reply:
> it would require a change to opencl-extensions/cl_intel_global_variable_access.asciidoc

Either that or we could handle it in OpenCL plugin, i.e. combine the two PI calls into a single OpenCL (and only give guarantee that variable's pointer is valid to use in read/write calls, but not necessarily with anything else, backend dependent).

> The offset is into the device variable, i.e. the operation will copy Count bytes to/from the device variable at byte-offset Offset.

This needs to be documented. I am still curious how would a caller know how many bytes are there, and how many is enough to read, and how to use them?

Reply:
> Either that or we could handle it in OpenCL plugin, i.e. combine the two PI calls into a single OpenCL (and only give guarantee that variable's pointer is valid to use in read/write calls, but not necessarily with anything else, backend dependent).

What is the motivation to expose a pointer?  It seems like the PI interface should provide an abstraction that works with all the existing backends and is sufficient to implement the corresponding SYCL API.  The SYCL API doesn't require a pointer to be exposed, and the OpenCL backend doesn't expose a pointer.  Why do we want to synthesize a pointer just so it can be exposed internally in the PI layer?

Reply:
Few consideration in no particular order/importance:

1) PI is transforming into standalone API that is usable outside of SYCL too, and other runtimes might want to use the variable's pointer for something other than read/write (prefetch, cmpxchg, etc)
2) It reduces the overall API size (sort of CISC vs. RISC architecture)
3) Get consistency with the other extension for getting function pointer: https://github.com/intel/llvm/blob/433a073e0b740451f92126f1b3c6075784ee2a93/sycl/include/CL/sycl/detail/pi.h#L1052

Reply:
There was a lot of discussion about whether the OpenCL API should expose a pointer for a "module scope variable", and it was a conscious decision not to do this.  I'd rather not reopen that discussion.  Even if we did expose a pointer, we certainly wouldn't be able to prefetch or cmpxchg to it from the host using regular pointer operations.  We'd still need to invent some API to schedule these operations from the host, and that API could just as easily take the name of the variable.

Reply:
We already have a prefetch API: https://github.com/intel/llvm/blob/433a073e0b740451f92126f1b3c6075784ee2a93/sycl/include/CL/sycl/detail/pi.h#L1734

I don't feel strongly about this, and so if you do, I am OK to go as is (but document the semantics).
We will likely repeat this discussion when implementing the extension  :)

Reply:
> This needs to be documented.

I agree. I have it documented in the upcoming patch as part of the definition. Would you like it documented here as well?

> I am still curious how would a caller know how many bytes are there, and how many is enough to read, and how to use them?

Specifically for `device_global` the runtime does not need to know because it knows the size of the type encapsulated by the `device_global`, which should be the same as the size of the device variable.

Reply:
> Would you like it documented here as well?

Yes, please add a short summary here as well since this is a design document.

Reply:
I have added a short summary of the expected semantics. @smaslov-intel - Is this what you had in mind?